src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,66 +191,182 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
         } else {
             mh = target;
         }
-        ClassLoader proxyLoader = intfc.getClassLoader();
-        if (proxyLoader == null) {
-            ClassLoader cl = Thread.currentThread().getContextClassLoader(); // avoid use of BCP
-            proxyLoader = cl != null ? cl : ClassLoader.getSystemClassLoader();
+
+        // We will just define the class as hidden under the interface
+        var info = IMPLEMENTATION_INFOS.get(intfc);
+        MethodHandle[] mhs = new MethodHandle[info.types.length + 1];
+        mhs[0] = target;
+        for (int i = 1; i < mhs.length; i++) {
+            mhs[i] = mh.asType(info.types[i - 1]);
         }
-        final Method[] methods = getSingleNameMethods(intfc);
-        if (methods == null)
-            throw newIllegalArgumentException("not a single-method interface", intfc.getName());
-        final MethodHandle[] vaTargets = new MethodHandle[methods.length];
-        for (int i = 0; i < methods.length; i++) {
-            Method sm = methods[i];
-            MethodType smMT = MethodType.methodType(sm.getReturnType(), sm.getParameterTypes());
-            MethodHandle checkTarget = mh.asType(smMT);  // make throw WMT
-            checkTarget = checkTarget.asType(checkTarget.type().changeReturnType(Object.class));
-            vaTargets[i] = checkTarget.asSpreader(Object[].class, smMT.parameterCount());
+
+        Object proxy;
+        try {
+            proxy = (Object) info.ctor.invokeExact(mhs); // non-varargs
+        } catch (Throwable e) {
```

Comment:
At least Error should be propagated (especially OutOfMemoryError)

Reply:
Done. I just noticed LambdaMetafactory has the same problem at no-arg constructor call site, might be worthy to create a bug as well.